docs/PerfEventArray.md

```diff
@@ -0,0 +1,147 @@
+      - Using `perf_buffer__new_raw` a consumer can attach to specific CPUs
+  2. Ring buffer maps support reserve and submit to separately allocate and then fill in the record.
+  3. For supported program types with a payload, perf event arrays can copy payload from the bpf context by
+  putting the length to copy in the `BPF_F_CTXLEN_MASK` field of the flags.
+      - `perf_event_output` takes the bpf context as an argument and the helper implementation copies the payload.
+          - The payload is whatever the data pointer of the program context points to (e.g. packet data including headers). The payload does not include the bpf program context structure itself.
+
+The main motivation for this proposal is to efficiently support payload capture from the context in bpf programs.
+- Supporting ring buffer reserve and submit in ebpf-for-windows is currently blocked on verifier support [#273](https://github.com/vbpf/ebpf-verifier/issues/273).
+- Without reserve+submit, using `ringbuf_output` for payload capture requires using a per-CPU array as scratch space to append the payload to the event before calling ringbuf_output.
+- The CTXLEN field in the flags of `perf_event_output` tells the kernel to append bytes from the payload to the record, avoiding the extra copy.
+  - On linux this works for specifc program types, on Windows this will work for any program type with a data pointer in the context.
+
+
+# Proposal
+
+The proposed behaviour matches linux, but currently only supports user-space consumers and bpf-program producers with a subset of the features.
+
+The plan is to implement perf buffers using the existing per-CPU and ring buffer map support in ebpf-for-windows.
+
+To match linux behaviour, by default the callback will only be called inside calls to `perf_buffer__poll()`.
+If the PERFBUF_FLAG_AUTO_CALLBACK flag is set, the callback will be automatically invoked when there is data available.
+
+1. Implement a new map type `BPF_MAP_TYPE_PERF_EVENT_ARRAY`.
+    1. Linux-compatible default behaviour.
+        - With linux-compatible behaviour and bpf interfaces, additional features from linux should be possible to add in the future.
+    2. Only support the perf ringbuffer (not other linux perf features).
+        - Only support bpf program producers with a single user-space consumer per event array.
+        - Features not supported include perf counters, hardware-generated perf events,
+          attaching bpf programs to perf events, and sending events from user-space to bpf programs.
+    3. In addition to the linux behaviour, automatically invoke the callback if the auto callback flag is set.
+2. Implement `perf_event_output` bpf helper function.
+    1. Only support writing to the current CPU (matches current linux restrictions).
+        - Specify current CPU in flags using BPF_F_INDEX_MASK or pass BPF_F_CURRENT_CPU.
+    2. Support BPF_F_CTXLEN_MASK flags any bpf program types with a data pointer in the context.
```

Comment:
What happens if BPF_F_CTXLEN_MASK flag is passed but the program type doesn't ahve a data pointer in the context? Should we define the error handling in this proposal?

(How does linux handle this? We should ideally align with that behavior)